kubernetes/cache/kube_cache.go

```diff
@@ -507,6 +531,39 @@ func (c *kubeCache) createKubernetesInformers(namespace string) informers.Shared
 
 	sharedInformers := informers.NewSharedInformerFactoryWithOptions(c.client.Kube(), c.refreshDuration, opts...)
 
+	// If this is a namespace-scoped cache, set a watch error handler on all the things we are watching.
+	// If the error is due to the client being forbidden from seeing the resource, this likely means the namespace
+	// has been deleted. In this case, we want to disable this namespace-scoped cache since it will not work for any
+	// resource. We add a watch handler on all informers because we don't know which one will be used first after
+	// the namespace deletion and we want to shut the informers down as quickly as we can.
+	if namespace != "" {
+		informersToWatchList := []cache.SharedIndexInformer{
+			sharedInformers.Apps().V1().Deployments().Informer(),
+			sharedInformers.Apps().V1().StatefulSets().Informer(),
+			sharedInformers.Apps().V1().DaemonSets().Informer(),
+			sharedInformers.Core().V1().Services().Informer(),
+			sharedInformers.Core().V1().Endpoints().Informer(),
+			sharedInformers.Core().V1().Pods().Informer(),
+			sharedInformers.Apps().V1().ReplicaSets().Informer(),
+			sharedInformers.Core().V1().ConfigMaps().Informer(),
+		}
+
+		watchErrorHandler := func(r *cache.Reflector, err error) {
+			if apierrors.IsForbidden(err) || apierrors.IsUnauthorized(err) {
```

Comment:
We're only checking for these two errors because I'm guessing the operator deletes the role?